moveit_core/collision_detection_fcl/src/collision_common.cpp

```diff
@@ -463,37 +462,141 @@ bool distanceCallback(fcl::CollisionObject* o1, fcl::CollisionObject* o2, void*
 
   if (always_allow_collision)
   {
-    min_dist = cdata->res_->distance;
-    return cdata->done_;
+    return false;
   }
+  if (cdata->req->verbose)
+    CONSOLE_BRIDGE_logDebug("Actually checking collisions between %s and %s", cd1->getID().c_str(),
+                            cd2->getID().c_str());
 
-  fcl::DistanceResult dist_result;
-  dist_result.update(cdata->res_->distance, nullptr, nullptr, fcl::DistanceResult::NONE,
-                     fcl::DistanceResult::NONE);  // can be faster
-  const double d = fcl::distance(o1, o2, fcl::DistanceRequest(), dist_result);
+  fcl::DistanceResult fcl_result;
+  DistanceResultsData dist_result;
+  double dist_threshold = cdata->req->distance_threshold;
+
+  const std::pair<std::string, std::string>& pc = cd1->getID() < cd2->getID() ?
+                                                        std::make_pair(cd1->getID(), cd2->getID()) :
+                                                        std::make_pair(cd2->getID(), cd1->getID());
+
+  DistanceMap::iterator it = cdata->res->distances.find(pc);
 
-  if (cdata->req_->verbose)
-    CONSOLE_BRIDGE_logDebug("Distance between %s and %s: %f", cd1->getID().c_str(), cd2->getID().c_str(), d);
 
-  if (d < 0)  // a penetration was found, no further distance calculations are necessary
+  if (it != cdata->res->distances.end())
   {
-    cdata->done_ = true;
-    cdata->res_->distance = -1;
+    if (cdata->req->type == DistanceRequestType::LIMITED)
+    {
+      // If at the limit for a given pair just return
+      if (it->second.size() >= cdata->req->max_contacts_per_body)
+      {
+        return cdata->done;
+      }
+    }
+    else if (cdata->req->type != DistanceRequestType::GLOBAL)
```

Comment:
These if-checks do not make sense to me. Shouldn't they equality instead of inequality checks?